File: desktop/EatfitDesktop/ViewModels/ProfileViewModel.cs
using Eatfit.V1;
using EatfitDesktop.Helpers;
using EatfitDesktop.Services;
using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Input;

namespace EatfitDesktop.ViewModels
{
    public class ProfileViewModel : INotifyPropertyChanged
    {
        private readonly GrpcProfileService _profileService;
        private readonly SessionService _sessionService;

        public ProfileViewModel(GrpcProfileService profileService, SessionService sessionService)
        {
            _profileService = profileService;
            _sessionService = sessionService;

            UpdateWeightCommand = new RelayCommand(async _ => await UpdateWeightAsync());
            UpdateNormsCommand = new RelayCommand(async _ => await UpdateNormsAsync());
        }

        private string _displayName = "Loading...";
        public string DisplayName
        {
            get => _displayName;
            set { _displayName = value; OnPropertyChanged(); }
        }

        private string _bio = "";
        public string Bio
        {
            get => _bio;
            set { _bio = value; OnPropertyChanged(); }
        }

        private string _height = "—";
        public string Height
        {
            get => _height;
            set { _height = value; OnPropertyChanged(); }
        }

        private string _weight = "—";
        public string Weight
        {
            get => _weight;
            set { _weight = value; OnPropertyChanged(); }
        }

        private string _age = "—";
        public string Age
        {
            get => _age;
            set { _age = value; OnPropertyChanged(); }
        }

        private string _gender = "—";
        public string Gender
        {
            get => _gender;
            set { _gender = value; OnPropertyChanged(); }
        }

        private string _birthDate = "—";
        public string BirthDate
        {
            get => _birthDate;
            set { _birthDate = value; OnPropertyChanged(); }
        }

        // Нормы
        private string _calories = "—";
        public string Calories
        {
            get => _calories;
            set { _calories = value; OnPropertyChanged(); }
        }

        private string _proteins = "—";
        public string Proteins
        {
            get => _proteins;
            set { _proteins = value; OnPropertyChanged(); }
        }

        private string _fats = "—";
        public string Fats
        {
            get => _fats;
            set { _fats = value; OnPropertyChanged(); }
        }

        private string _carbs = "—";
        public string Carbs
        {
            get => _carbs;
            set { _carbs = value; OnPropertyChanged(); }
        }

        private string _water = "—";
        public string Water
        {
            get => _water;
            set { _water = value; OnPropertyChanged(); }
        }

        public ICommand UpdateWeightCommand { get; }
        public ICommand UpdateNormsCommand { get; }

        public async Task LoadProfileAsync()
        {
            var userId = _sessionService.UserId;
            if (string.IsNullOrEmpty(userId)) return;

            var profile = await _profileService.GetProfileAsync(userId);
            if (profile == null) return;

            // Имя
            if (!string.IsNullOrEmpty(profile.FirstName) || !string.IsNullOrEmpty(profile.LastName))
            {
                DisplayName = $"{profile.FirstName} {profile.LastName}".Trim();
            }
            else
            {
                DisplayName = profile.Login;
            }

            Bio = profile.Bio;
            Height = profile.Height > 0 ? $"{profile.Height} cm" : "—";
            Weight = profile.CurrentWeight > 0 ? $"{profile.CurrentWeight} kg" : "—";
            Age = profile.Age > 0 ? profile.Age.ToString() : "—";
            Gender = profile.Gender switch
            {
                "male" => "Male",
                "female" => "Female",
                "other" => "Other",
                _ => "—"
            };
            BirthDate = !string.IsNullOrEmpty(profile.BirthDate) ? profile.BirthDate : "—";

            // Нормы
            if (profile.Norms != null)
            {
                Calories = $"{profile.Norms.Calories:F0} kcal";
                Proteins = $"{profile.Norms.Proteins:F0} g";
                Fats = $"{profile.Norms.Fats:F0} g";
                Carbs = $"{profile.Norms.Carbs:F0} g";
                Water = $"{profile.Norms.Water:F0} ml";
            }
        }

        private async Task UpdateWeightAsync()
        {
            var input = Microsoft.VisualBasic.Interaction.InputBox(
                "Enter new weight (kg):", "Update Weight", "70");

            if (double.TryParse(input, out double weight) && weight > 0)
            {
                var userId = _sessionService.UserId;
                if (string.IsNullOrEmpty(userId)) return;

                var success = await _profileService.UpdateWeightAsync(userId, weight);
                if (success)
                {
                    Weight = $"{weight} kg";
                    MessageBox.Show("Weight updated!", "Success", MessageBoxButton.OK, MessageBoxImage.Information);
                }
                else
                {
                    MessageBox.Show("Failed to update weight", "Error", MessageBoxButton.OK, MessageBoxImage.Error);
                }
            }
        }

        private async Task UpdateNormsAsync()
        {
            // Создаём простое диалоговое окно для ввода норм
            var dialog = new Views.NormsDialog();
            if (dialog.ShowDialog() == true)
            {
                var userId = _sessionService.UserId;
                if (string.IsNullOrEmpty(userId)) return;

                var success = await _profileService.UpdateNormsAsync(
                    userId,
                    dialog.Calories,
                    dialog.Proteins,
                    dialog.Fats,
                    dialog.Carbs,
                    dialog.Water
                );

                if (success)
                {
                    await LoadProfileAsync();
                    MessageBox.Show("Norms updated!", "Success", MessageBoxButton.OK, MessageBoxImage.Information);
                }
            }
        }

        public event PropertyChangedEventHandler? PropertyChanged;
        protected void OnPropertyChanged([CallerMemberName] string? propertyName = null)
            => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }
}

File: android/app/src/main/java/enzio/android/eatfit/ui/screens/profile/ProfileViewModel.kt
package enzio.android.eatfit.ui.screens.profile

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import enzio.android.eatfit.data.local.SessionManager
import enzio.android.eatfit.domain.ProfileRepository
import enzio.android.eatfit.proto.ProfileData
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

data class ProfileState(
    val displayName: String = "Loading...",
    val login: String = "",
    val bio: String = "",
    val height: String = "—",
    val weight: String = "—",
    val age: String = "—",
    val gender: String = "—",
    val birthDate: String = "—",
    val calories: String = "—",
    val proteins: String = "—",
    val fats: String = "—",
    val carbs: String = "—",
    val water: String = "—",
    val isLoading: Boolean = true
)

class ProfileViewModel(
    private val profileRepository: ProfileRepository,
    private val sessionManager: SessionManager
) : ViewModel() {

    private val _state = MutableStateFlow(ProfileState())
    val state: StateFlow<ProfileState> = _state

    init {
        loadProfile()
    }

    fun loadProfile() {
        viewModelScope.launch {
            val userId = sessionManager.userId ?: ""
            if (userId.isEmpty()) return@launch

            val profile = profileRepository.getProfile(userId)
            if (profile != null) {
                _state.value = _state.value.copy(
                    displayName = profile.displayName(),
                    login = profile.login,
                    bio = profile.bio,
                    height = if (profile.height > 0) "${profile.height} cm" else "—",
                    weight = if (profile.currentWeight > 0) "${profile.currentWeight} kg" else "—",
                    age = if (profile.age > 0) profile.age.toString() else "—",
                    gender = when (profile.gender) {
                        "male" -> "Male"
                        "female" -> "Female"
                        "other" -> "Other"
                        else -> "—"
                    },
                    birthDate = profile.birthDate.ifEmpty { "—" },
                    calories = if (profile.hasNorms()) "${profile.norms.calories.toInt()} kcal" else "—",
                    proteins = if (profile.hasNorms()) "${profile.norms.proteins.toInt()} g" else "—",
                    fats = if (profile.hasNorms()) "${profile.norms.fats.toInt()} g" else "—",
                    carbs = if (profile.hasNorms()) "${profile.norms.carbs.toInt()} g" else "—",
                    water = if (profile.hasNorms()) "${profile.norms.water.toInt()} ml" else "—",
                    isLoading = false
                )
            } else {
                _state.value = _state.value.copy(isLoading = false)
            }
        }
    }

    fun updateWeight(weight: Double) {
        viewModelScope.launch {
            val userId = sessionManager.userId ?: return@launch
            if (profileRepository.setWeight(userId, weight)) {
                // Перезагружаем профиль, чтобы вес обновился
                loadProfile()
            }
        }
    }

    fun updateNorms(calories: Double, proteins: Double, fats: Double, carbs: Double, water: Double) {
        viewModelScope.launch {
            val userId = sessionManager.userId ?: return@launch
            if (profileRepository.setNorms(userId, calories, proteins, fats, carbs, water)) {
                loadProfile()
            }
        }
    }

    private fun ProfileData.displayName(): String {
        return if (!firstName.isNullOrEmpty() || !lastName.isNullOrEmpty()) {
            "${firstName ?: ""} ${lastName ?: ""}".trim()
        } else {
            login
        }
    }
}